File: contracts/src/main/java/com/densitech/largescale/contracts/AuthService.kt
package com.densitech.largescale.contracts

import kotlinx.coroutines.flow.StateFlow

/**
 * Authentication service — single source of truth for user identity.
 *
 * Lives in `:contracts` so feature modules can inject it directly
 * without depending on `:core`.
 *
 * Wire Core reacts to [UserAuthenticatedEvent] / [UserLoggedOutEvent] published
 * here to update [ModuleContext.currentRole] automatically.
 *
 * Usage in a ViewModel:
 * ```
 * @HiltViewModel
 * class ProfileViewModel @Inject constructor(
 *     private val authService: AuthService
 * ) : ViewModel() {
 *     val user = authService.currentUser
 * }
 * ```
 */
interface AuthService {

    /** Currently authenticated user, or null when logged out. */
    val currentUser: StateFlow<User?>

    /** Current role — mirrors [currentUser]?.role, defaults to [Role.GUEST]. */
    val currentRole: StateFlow<Role>

    /** Convenience flag: true when a user is signed in. */
    val isAuthenticated: StateFlow<Boolean>

    /**
     * Attempt to authenticate with [username] and [password].
     *
     * On success: persists the session, publishes [UserAuthenticatedEvent], returns [Result.success].
     * On failure: returns [Result.failure] with a descriptive exception — caller shows error UI.
     */
    suspend fun login(username: String, password: String): Result<User>

    /**
     * Sign out the current user.
     * Clears the session, resets role to [Role.GUEST], publishes [UserLoggedOutEvent].
     */
    suspend fun logout()

    /**
     * Restore a previously persisted session (called on app start).
     * Returns the restored [User] or null if no valid session exists.
     */
    suspend fun restoreSession(): User?
}

File: flutter/contracts/lib/src/auth_service.dart
import 'user.dart';
import 'role.dart';

/// Authentication service — single source of truth for user identity.
///
/// Lives in [contracts] so feature modules can inject it without depending on [core].
/// Mirrors Android [AuthService] interface.
///
/// Wire Core reacts to [UserAuthenticatedEvent]/[UserLoggedOutEvent] published
/// here to update [ModuleContext.currentRole] automatically.
abstract class AuthService {
  /// Currently authenticated user, or null when logged out.
  User? get currentUserValue;

  /// Current role — mirrors [currentUserValue]?.role, defaults to [Role.GUEST].
  Role get currentRole;

  /// True when a user is signed in.
  bool get isAuthenticated;

  /// Stream of role changes — used by [ModuleContext] to react to auth events.
  Stream<Role> get currentRoleStream;

  /// Attempt to authenticate with [username] and [password].
  ///
  /// Returns the authenticated [User] on success.
  /// Throws with a descriptive message on failure.
  Future<User> login(String username, String password);

  /// Sign out the current user.
  /// Resets role to [Role.GUEST] and publishes [UserLoggedOutEvent].
  Future<void> logout();

  /// Restore a previously persisted session (called on app start).
  /// Returns the restored [User] or null if no valid session exists.
  Future<User?> restoreSession();
}
